library/alloc/src/slice.rs

```diff
@@ -159,64 +160,62 @@ mod hack {
         where
             Self: Sized;
     }
 
-    impl<T: Clone> ConvertVec for T {
+    impl<T: Clone> ConvertBoxed for T {
         #[inline]
-        default fn to_vec<A: Allocator>(s: &[Self], alloc: A) -> Vec<Self, A> {
-            struct DropGuard<'a, T, A: Allocator> {
-                vec: &'a mut Vec<T, A>,
+        default fn to_boxed_slice<A: Allocator>(s: &[Self], alloc: A) -> Box<[Self], A> {
+            struct DropGuard<T> {
+                dst: *mut T,
                 num_init: usize,
             }
-            impl<'a, T, A: Allocator> Drop for DropGuard<'a, T, A> {
+            impl<T> Drop for DropGuard<T> {
                 #[inline]
                 fn drop(&mut self) {
-                    // SAFETY:
-                    // items were marked initialized in the loop below
+                    let init_slice = core::ptr::slice_from_raw_parts_mut(self.dst, self.num_init);
+                    // SAFETY: all elements in this slice have been initialized
                     unsafe {
-                        self.vec.set_len(self.num_init);
+                        core::ptr::drop_in_place(init_slice);
                     }
                 }
             }
-            let mut vec = Vec::with_capacity_in(s.len(), alloc);
-            let mut guard = DropGuard { vec: &mut vec, num_init: 0 };
-            let slots = guard.vec.spare_capacity_mut();
-            // .take(slots.len()) is necessary for LLVM to remove bounds checks
-            // and has better codegen than zip.
-            for (i, b) in s.iter().enumerate().take(slots.len()) {
-                guard.num_init = i;
-                slots[i].write(b.clone());
+
+            let mut boxed = Box::new_uninit_slice_in(s.len(), alloc);
+            let mut guard =
+                DropGuard { dst: MaybeUninit::slice_as_mut_ptr(&mut boxed), num_init: 0 };
+            while guard.num_init < s.len() {
+                boxed[guard.num_init].write(s[guard.num_init].clone());
+                guard.num_init += 1;
             }
             core::mem::forget(guard);
-            // SAFETY:
-            // the vec was allocated and initialized above to at least this length.
-            unsafe {
-                vec.set_len(s.len());
-            }
-            vec
+            // SAFETY: each element is initialized by the loop above
+            unsafe { boxed.assume_init() }
         }
     }
 
-    impl<T: Copy> ConvertVec for T {
+    impl<T: Copy> ConvertBoxed for T {
         #[inline]
-        fn to_vec<A: Allocator>(s: &[Self], alloc: A) -> Vec<Self, A> {
-            let mut v = Vec::with_capacity_in(s.len(), alloc);
-            // SAFETY:
-            // allocated above with the capacity of `s`, and initialize to `s.len()` in
-            // ptr::copy_to_non_overlapping below.
+        fn to_boxed_slice<A: Allocator>(s: &[Self], alloc: A) -> Box<[Self], A> {
+            let mut boxed = Box::new_uninit_slice_in(s.len(), alloc);
```

Comment:
Lots of effort went into the Vec allocation path to reduce the amount of IR generated, perhaps the Box equivalent simply isn't as fine-tuned and that causes the regressions?

Reply:
They're identical except for the additional `RawVec::into_box()` call, which just moves some pointers and lengths around, so it seems unlikely that this would be the culprit:
```rust
impl<T, A: Allocator> Box<[T], A> {
    pub fn new_uninit_slice_in(len: usize, alloc: A) -> Box<[mem::MaybeUninit<T>], A> {
        unsafe { RawVec::with_capacity_in(len, alloc).into_box(len) }
    }
}

impl<T, A: Allocator> Vec<T, A> {
    pub fn with_capacity_in(capacity: usize, alloc: A) -> Self {
        Vec { buf: RawVec::with_capacity_in(capacity, alloc), len: 0 }
    }
}

impl<T, A: Allocator> RawVec<T, A> {
    pub unsafe fn into_box(self, len: usize) -> Box<[MaybeUninit<T>], A> {
        // Sanity-check one half of the safety requirement (we cannot check the other half).
        debug_assert!(
            len <= self.capacity(),
            "`len` must be smaller than or equal to `self.capacity()`"
        );

        let me = ManuallyDrop::new(self);
        unsafe {
            let slice = slice::from_raw_parts_mut(me.ptr() as *mut MaybeUninit<T>, len);
            Box::from_raw_in(slice, ptr::read(&me.alloc))
        }
    }
}
```